[Phua Guan Wei] iP

GWphua opened the pull request:
### Virushade

> “Your mind is for having ideas, not holding them.” – David Allen ! 
> [(source)](https://dansilvestre.com/productivity-quotes/)

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* *easy* to learn
* ~~FAST~~ **SUPER** FAST to use.

All you need to do is:
1. Add tasks to Virushade.
2. Let Virushade manage the tasks for you. 😀

Features:
- [x] Add / Delete tasks
- [x] List all of your tasks.
- [x] Search tasks.
- [x] Indication of overdue tasks. (coming soon)



If you are a Java programmer, you can use it to practice your code too! Here's the class `Virushade`:
```java
package virushade;

import virushade.tasks.TaskList;

/**
 * The Main class of the application.
 */
public class Virushade {
    private Storage storageFile;
    private TaskList taskList;

    /**
     * The constructor for Virushade.
     * @param fileName The file path of the file where data is written to.
     */
    public Virushade(String fileName) {
        storageFile = new Storage(fileName);
        try {
            TaskList taskList = new TaskList(storageFile);
        } catch (VirushadeException e) {
            Ui.handleVirushadeException(e);
        }
    }

    /**
     * The running sequence for Virushade.
     */
    public void run() {
        Ui.interact();
    }

    /**
     * Runs our program under "data/Virushade.txt".
     * @param args Not used.
     */
    public static void main(String[] args) {
        new Virushade("data/Virushade.txt").run();
    }
}
```

Commit:
Add Gradle support

GWphua committed:
Added Greet, Echo, and Exit to Duke.java

GWphua committed:
Changed functionalities, renamed Duke to Virushade.

GWphua committed:
Added Add and List functions under Memory.

GWphua committed:
Renamed Memory to TaskList, added Task and its functionalities.
Every string in TaskList (or previously Memory) is now a Task.

GWphua committed:
Added "Mark as Done" functionality in Virushade.

GWphua committed:
Added ToDos, Events, Deadlines in Virushade.

GWphua committed:
Added automated testing using the I/O redirection technique.
Note input file (input.exe) and testcase file (EXPECTED.TXT)

GWphua committed:
Added error-handling, also updated input.txt and EXPECTED.TXT to cater for cases where none of the 3 types of tasks are created.

GWphua committed:
Added Delete Function

GWphua committed:
Add "Save" functionality.

GWphua committed:
Add ability to read Dates and Times

GWphua committed:
Revert "Add "Save" functionality."

This reverts commit 9eb7bb1c7977bead6e3a3b331234b9325f6701ec.

GWphua committed:
Merge branch 'branch-Level-8'

GWphua committed:
Revert "Revert "Add "Save" functionality.""

This reverts commit bf6b0b5e68dffb1df9d7e6589d34a74ce1a8e750.

GWphua committed:
Merge branch 'master' into branch-Level-7

GWphua committed:
Abstract out "Storage" from TaskList.

GWphua committed:
Merge branch 'branch-Level-7'

GWphua committed:
Refactor code to add more OOP.

GWphua committed:
Added packages.

GWphua committed:
Add JUnit Tests

GWphua committed:
Created JAR Files.

GWphua committed:
Add Java Docs.

GWphua committed:
Refactor all according to Java Coding Standards

GWphua committed:
Implement "Find" functionality

GWphua committed:
Merge branch 'branch-A-JavaDoc'

GWphua committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/virushade/Storage.java
#	src/main/java/virushade/VirushadeException.java

GWphua committed:
Merge branch 'branch-Level-9'

GWphua committed:
Bug fixes to delete function.

SKCheongBrian commented on src/main/java/virushade/tasks/TaskList.java:
```diff
@@ -0,0 +1,210 @@
+            listCount++;
+            System.out.println("Added: " + addedTask.getTaskDescription());
+            System.out.printf("Now you have %d tasks in the list.\n", listCount);
+            updateFile(generateList());
```
Good idea to update the file everytime a task is added. That way your tasks will still be saved for recovery in case the application crashs.

SKCheongBrian commented on src/main/java/virushade/Ui.java:
```diff
@@ -0,0 +1,152 @@
+     * @param str The input.
+     * @throws VirushadeException The response error message for the user.
+     */
+    private static void handleUnexpectedInputs(String str) throws VirushadeException {
+        if (str.startsWith("todo")) {
+            throw new VirushadeException("OOPS!!! The description of a todo task cannot be empty.");
+        } else if (str.startsWith("deadline")) {
+            throw new VirushadeException("OOPS!!! The description of a deadline task cannot be empty.");
+        } else if (str.startsWith("event")) {
+            throw new VirushadeException("OOPS!!! The description of an event task cannot be empty.");
+        } else if (str.startsWith("done")) {
+            throw new VirushadeException("OOPS!!! Please enter an integer after 'done'.");
+        } else if (str.startsWith("delete")) {
+            throw new VirushadeException("OOPS!!! Please enter an integer after 'delete'.");
+        } else if (str.startsWith("find")) {
+            throw new VirushadeException("OOPS!!! Please tell me what to find.");
+        } else {
+            throw new VirushadeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
+    }
```
You can consider using a switch statement here.

GWphua replied:
Thanks for the input!!

SKCheongBrian commented on src/main/java/virushade/Ui.java:
```diff
@@ -0,0 +1,152 @@
+     * @param str The input.
+     * @throws VirushadeException The response error message for the user.
+     */
+    private static void handleUnexpectedInputs(String str) throws VirushadeException {
+        if (str.startsWith("todo")) {
+            throw new VirushadeException("OOPS!!! The description of a todo task cannot be empty.");
+        } else if (str.startsWith("deadline")) {
+            throw new VirushadeException("OOPS!!! The description of a deadline task cannot be empty.");
+        } else if (str.startsWith("event")) {
+            throw new VirushadeException("OOPS!!! The description of an event task cannot be empty.");
+        } else if (str.startsWith("done")) {
+            throw new VirushadeException("OOPS!!! Please enter an integer after 'done'.");
+        } else if (str.startsWith("delete")) {
+            throw new VirushadeException("OOPS!!! Please enter an integer after 'delete'.");
+        } else if (str.startsWith("find")) {
+            throw new VirushadeException("OOPS!!! Please tell me what to find.");
+        } else {
+            throw new VirushadeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
+    }
```
You can consider using a switch statement here.

SKCheongBrian commented on src/main/java/virushade/Ui.java:
```diff
@@ -0,0 +1,152 @@
+     */
+    private static void assignTask(String str) {
+        try {
+            if (str.equals("list")) {
+                TaskList.list();
+            } else if (str.startsWith("find ")) {
+                TaskList.find(str.substring(5));
+            } else if (str.startsWith("done ")) {
+                TaskList.completeTask(str.substring(5));
+            } else if (str.startsWith("todo ")) {
+                TaskList.add(str.substring(5), "TODO");
+            } else if (str.startsWith("deadline ")) {
+                TaskList.add(str.substring(9), "DEADLINE");
+            } else if (str.startsWith("event ")) {
+                TaskList.add(str.substring(6), "EVENT");
+            } else if (str.startsWith("delete ")) {
+                TaskList.delete(str.substring(7));
+            } else {
+                handleUnexpectedInputs(str);
+            }
```
You can consider using a switch statement here.

SKCheongBrian commented on src/main/java/virushade/Virushade.java:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * The running sequence for Virushade.
+     */
+    public void run() {
+        Ui.interact();
+    }
```
The Ui should not be responsible for all the backend work being done, it may be better to just focus the Ui on printing on the console and getting input from the console. Perhaps another class can be made to communicate between Ui, TaskList and Storage.

SKCheongBrian reviewed (commented):
Your code seems very well done.

nniiggeell commented on src/main/java/virushade/Storage.java:
```diff
@@ -0,0 +1,88 @@
+
+    /**
+     * Reads from STORAGE_FILE and updates the given ArrayList input according to the lines in STORAGE_FILE.
+     * @param tasks The ArrayList input to update information from STORAGE_FILE to.
```
Should there be an empty line separating description and @param? Similar for all javadoc

nniiggeell commented on src/main/java/virushade/tasks/TaskList.java:
```diff
@@ -0,0 +1,210 @@
+    /**
+     * This is the array that Virushade keeps the tasks in.
+     */
+    private static final ArrayList<Task> TASKS = new ArrayList<>();
```
Should this be final? Will TASKS ever change? Sorry if noob question it seems confusing to me

GWphua replied:
I also not sure haha, I made it final cuz IntelliJ hinted at me. I'm not sure what a final ArrayList means though.

nniiggeell reviewed (commented):
Very minimal coding style errors!
Apart from the javadoc comments I cant spot any more errors. Well done!

GWphua committed:
Merge branch 'add-gradle-support'

# Conflicts:
#	text-ui-test/runtest.sh

GWphua committed:
Implement JavaFX Hello World

GWphua committed:
Create GUI for Virushade.

GWphua committed:
Add echo user function.

GWphua committed:
Complete implementation of GUI. To work on getResponse.

GWphua committed:
Implement Level 10

GWphua committed:
Merge branch 'branch-Level-10'

GWphua committed:
Clean up code, implemented delay after saying 'bye'.

GWphua committed:
Add Assert, tidied up code to adhere to Java Style Guide.

GWphua committed:
Add checkstyle.

GWphua committed:
Merge pull request #1 from GWphua/branch-A-Assertions

Branch-A-Assertions

GWphua committed:
Refactor code.

Refactor the code according to achieve better code quality.
Extracted out most long methods, and implemented SLAP for methods.

GWphua committed:
Update header comments.

Updated all header comments to appear more consistent, and in-line with the coding standard.

GWphua committed:
Merge pull request #3 from GWphua/branch-A-CodeQuality

Branch Code Quality

GWphua committed:
Add sort class. Yet to implement.

GWphua committed:
Add sort

Sort function for now has sort name and sort task. Sort name sorts the tasks alphanumerically. Sort task sorts the tasks by the task type (e.g. Todo/Event/Deadline). Sort Task works fine. Sort name still appears to be a bit buggy.

GWphua committed:
Bug fixes onto Sort. Now works correctly.

GWphua committed:
Merge pull request #4 from GWphua/branch-BCD-Extension

Branch Sort

GWphua committed:
Add new profile pictures.

GWphua committed:
Solve small bug that happens for StringManipulator.

GWphua committed:
Settle small issues with the strings and response of Virushade.

GWphua committed:
Write user guide.

GWphua committed:
Settle small coding convention adjustments.
